From da674e9cba597a4108361201890fb4ccd03d0f88 Mon Sep 17 00:00:00 2001
From: Sergii Nikolaiev <sergii.nikolaiev@lge.com>
Date: Thu, 28 May 2015 21:12:34 +0300
Subject: [PATCH 06/17] [PATCH 01/10] Adds noexec option for fork and spawn
 APIs of nodejs

:Release Notes:
Adds noexec option, which disables exec() syscall of
fork and spawn APIs of nodejs.

:Detailed Notes:
Initial Andrii Motsok`s patch:
https://gpro.lgsvl.com/#/c/54912/
sergii.nikolaiev@lge.com: fixed crashes in child process after fork
alexander.pryadka@lge.com: code refactoring. Added noexec option.

:Testing Performed:
Tested using fork server and patches:
https://g2g.lgsvl.com/#/c/7712
https://g2g.lgsvl.com/#/c/7881/
https://g2g.lgsvl.com/#/c/7883/
https://g2g.lgsvl.com/#/c/7884/
https://g2g.lgsvl.com/#/c/7885/
executing example helloworld, activityexample and some other
dynamic system services.

:QA Notes:

:Issues Addressed:
[DRD-4340] Using fork for loading new node instance

Change-Id: I66437f45c8134a0094ea2836e39ef9162ca0c058
Open-webOS-DCO-1.0-Signed-off-by: Sergii Nikolaiev <sergii.nikolaiev@lge.com>
Reviewed-on: https://gpro.lgsvl.com/111236
Reviewed-by: Sergii Nikolaiev <sergii.nikolaiev@lge.com>
Tested-by: Sergii Nikolaiev <sergii.nikolaiev@lge.com>
Reviewed-by: Ievgen Sherstobitov <ievgen.sherstobitov@lge.com>
Reviewed-by: Andrii Koval <andrii.koval@lge.com>
Reviewed-by: Alexander Pryadka <alexander.pryadka@lge.com>
Reviewed-by: Sergiy Kryvonos <sergiy.kryvonos@lge.com>
---
 deps/uv/include/uv.h          |  7 ++++++-
 deps/uv/src/unix/process.c    | 46 ++++++++++++++++++++++++++++++++++++++-----
 lib/child_process.js          |  3 ++-
 lib/internal/child_process.js |  4 ++++
 src/env.h                     |  1 +
 src/process_wrap.cc           |  6 ++++++
 6 files changed, 60 insertions(+), 7 deletions(-)

diff --git a/deps/uv/include/uv.h b/deps/uv/include/uv.h
index 1dd992e..0780474 100644
--- a/deps/uv/include/uv.h
+++ b/deps/uv/include/uv.h
@@ -939,7 +939,12 @@ enum uv_process_flags {
    * option is only meaningful on Windows systems. On Unix it is silently
    * ignored.
    */
-  UV_PROCESS_WINDOWS_HIDE = (1 << 4)
+  UV_PROCESS_WINDOWS_HIDE = (1 << 4),
+  /*
+   * Fork new process without exec system call. This option is only meaningful
+   * on Linux systems.
+   */
+  UV_PROCESS_NO_EXEC = (1 << 5)
 };
 
 /*
diff --git a/deps/uv/src/unix/process.c b/deps/uv/src/unix/process.c
index f2fe305..8dadc17 100644
--- a/deps/uv/src/unix/process.c
+++ b/deps/uv/src/unix/process.c
@@ -373,12 +373,36 @@ static void uv__process_child_init(const uv_process_options_t* options,
   }
 
   if (options->env != NULL) {
-    environ = options->env;
+    if (options->flags & UV_PROCESS_NO_EXEC) {
+      char **env_copy;
+      int envc = 0;
+      while (options->env[envc] != NULL) {
+        ++envc;
+      }
+      env_copy = (char**)malloc(sizeof(char*) * (envc + 1));
+      if (env_copy != NULL) {
+        int i = 0;
+        for (; i < envc; ++i) {
+          env_copy[i] = strdup(options->env[i]);
+          if (env_copy[i] == NULL) {
+            abort();
+          }
+        }
+        env_copy[envc] = NULL;
+        environ = env_copy;
+      } else {
+        abort();
+      }
+    } else {
+      environ = options->env;
+    }
   }
 
-  execvp(options->file, options->args);
-  uv__write_int(error_fd, -errno);
-  _exit(127);
+  if (!(options->flags & UV_PROCESS_NO_EXEC)) {
+    execvp(options->file, options->args);
+    uv__write_int(error_fd, -errno);
+    _exit(127);
+  }
 }
 #endif
 
@@ -405,7 +429,8 @@ int uv_spawn(uv_loop_t* loop,
                               UV_PROCESS_SETGID |
                               UV_PROCESS_SETUID |
                               UV_PROCESS_WINDOWS_HIDE |
-                              UV_PROCESS_WINDOWS_VERBATIM_ARGUMENTS)));
+                              UV_PROCESS_WINDOWS_VERBATIM_ARGUMENTS |
+                              UV_PROCESS_NO_EXEC)));
 
   uv__handle_init(loop, (uv_handle_t*)process, UV_PROCESS);
   QUEUE_INIT(&process->queue);
@@ -470,6 +495,17 @@ int uv_spawn(uv_loop_t* loop,
 
   if (pid == 0) {
     uv__process_child_init(options, stdio_count, pipes, signal_pipe[1]);
+    if (options->flags & UV_PROCESS_NO_EXEC) {
+      /* Release lock in child process */
+      uv_rwlock_wrunlock(&loop->cloexec_lock);
+      uv__close(signal_pipe[0]);
+      uv__close(signal_pipe[1]);
+
+      process->pid = pid;
+
+      free(pipes);
+      return 0;
+    }
     abort();
   }
 
diff --git a/lib/child_process.js b/lib/child_process.js
index 0e1d8f2..196b69e 100644
--- a/lib/child_process.js
+++ b/lib/child_process.js
@@ -375,7 +375,8 @@ var spawn = exports.spawn = function(/*file, args, options*/) {
     envPairs: opts.envPairs,
     stdio: options.stdio,
     uid: options.uid,
-    gid: options.gid
+    gid: options.gid,
+    noexec: options.noexec
   });
 
   return child;
diff --git a/lib/internal/child_process.js b/lib/internal/child_process.js
index 6033e97..3d2fc97 100644
--- a/lib/internal/child_process.js
+++ b/lib/internal/child_process.js
@@ -321,6 +321,10 @@ ChildProcess.prototype.spawn = function(options) {
 
   this.pid = this._handle.pid;
 
+  if (this.pid == 0) {
+    return 0;
+  }
+
   for (i = 0; i < stdio.length; i++) {
     const stream = stdio[i];
     if (stream.type === 'ignore') continue;
diff --git a/src/env.h b/src/env.h
index bb4dab0..60a3a89 100644
--- a/src/env.h
+++ b/src/env.h
@@ -222,6 +222,7 @@ namespace node {
   V(write_queue_size_string, "writeQueueSize")                                \
   V(x_forwarded_string, "x-forwarded-for")                                    \
   V(zero_return_string, "ZERO_RETURN")                                        \
+  V(noexec_string, "noexec")
 
 #define ENVIRONMENT_STRONG_PERSISTENT_PROPERTIES(V)                           \
   V(as_external, v8::External)                                                \
diff --git a/src/process_wrap.cc b/src/process_wrap.cc
index 3dcde09..f0ff7ce 100644
--- a/src/process_wrap.cc
+++ b/src/process_wrap.cc
@@ -151,6 +151,12 @@ class ProcessWrap : public HandleWrap {
       return env->ThrowTypeError("Bad argument");
     }
 
+    // options.noexec
+    Local<String> noexec_key = env->noexec_string();
+    if (js_options->Get(noexec_key)->IsTrue()) {
+        options.flags |= UV_PROCESS_NO_EXEC;
+    }
+
     // options.args
     Local<Value> argv_v = js_options->Get(env->args_string());
     if (!argv_v.IsEmpty() && argv_v->IsArray()) {
-- 
1.9.1

